### PR TITLE
[Codegen] Support inferring scalable vectors for remainder loops

### PR DESCRIPTION
**Pull request by MacDue**
Follow on for #17891 that extends the scalable vector size inference to work for remainder loops. The upper bound of a scalable remainder dim is an expression of the form `(vscale * n) + cst` (cst <= 0).

This patch adds some simple pattern matching that rounds up the upper bound by removing the `+ cst`, which allows matching the scalable quantity.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/test/generic_vectorization.mlir [hunk 1]
```diff
@@ -405,3 +405,43 @@ func.func @dynamic_fill_with_scalable_tiling_infer_vector_size(%arg0: tensor<1x6
...
+        scf.yield %inserted_slice : tensor<1x67x120x100xf32>
+      }
+      scf.yield %3 : tensor<1x67x120x100xf32>
+    }
+    scf.yield %1 : tensor<1x67x120x100xf32>
+  }
+  return %0 : tensor<1x67x120x100xf32>
+}
+
+// CHECK-MASK-LABEL: func.func @dynamic_fill_with_scalable_tiling_infer_remainder_vector_size
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Common/test/generic_vectorization.mlir:442]: There's a mask here, right? Also, what's the codegen today without this patch?

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 2]
```diff
@@ -1175,9 +1176,25 @@ computeDimUpperBound(Value shapedValue, unsigned dimNum,
...
-  return failure();
+  if (failed(maybeDimBound))
+    return failure();
+  auto boundSize = maybeDimBound->getSize();
+  if (succeeded(boundSize))
+    return boundSize;
+  if (roundUp == RoundUpVscaleMultiple::No)
+    return failure();
+  // If the upper bound map is of the form `add(subExpr, cst)` (cst <= 0),
+  // round it up to `subExpr`, which is _likely_ to be a vscale multiple.
```
Comment (banach-space) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp]: Which expression in your test would this apply to?

Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp]: It's not something you see in the test source. It's the expression computed by value bounds analysis for the scalable upper bound of a dim.

File: compiler/src/iree/compiler/Codegen/Common/test/generic_vectorization.mlir (hunk 1 above)
Comment (MacDue) [reply compiler/src/iree/compiler/Codegen/Common/test/generic_vectorization.mlir:442]: There's a mask (vector sizes are only inferred when masking is enabled). Without this patch it's not vectorized.

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp (hunk 2 above)
Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/Utils/Utils.cpp]: > , which is _likely_ to be a vscale multiple

I'd skip this - it implies that this logic is specific to `vscale` (while it isn't).

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp (hunk 2 above)
Comment (MacDue) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp]:
```suggestion
  // If the upper bound map is of the form `add(subExpr, cst)` (cst <= 0),
  // round it up to `subExpr` (and try matching the bound again).
```

**Pull request merged**